ls1intum/Artemis
# `Tutorial groups`: Improve readability of tutorial group table

### Checklist
#### General
#### Client

### Motivation and Context
@krusche requested some changes to the way tutorial groups are displayed in the table to increase usability / readability.


### Description
- Remove seconds from time display (e.g. 14:00:00 -> 14:00) in commit 36d00e596b7a5455fa40dec0ed78e8d5f90ae7cf
- Combine Time Weekday and Repetition pattern in one string (e.g. Every Week, Monday, 14:00-15:00) 070b15b6c621c512f6ed2a3e3bcbdf49779be4c3
- Only show repetition frequency if it is not weekly  c3a2b8b480282af47202dda09a2faf8d6698d711
- Show the average attendance as overlay to progress bar instead of new column eb773b316a3dd51da1075d903cfbe19a1e8a6f76
- Combine capacity and registered students in one column 095c6db76fccee630159c481575ea88ba2a2538b
- Only show language, campus, and online/offline columns if not all equal for all tutorial groups 1a31e0c4cad6be41fa26986493b5ef35c1d1b549

### Steps for Testing
- Verify that the time display does not show seconds for any tutorial group or event. For example, 14:00:00 should be displayed as 14:00.
- Verify that the time weekday and repetition pattern are shown in one string for each tutorial group or event. For example, Every Week, Monday, 14:00-15:00 should be displayed instead of separate columns for time, weekday and repetition frequency.
- Verify that the repetition frequency is hidden if it is weekly. For example, if the repetition pattern is Every Week, it should not be displayed at all in the column. It should be displayed if the group repeats every two weeks.
- Verify that the average attendance is shown as an overlay to the progress bar for each tutorial group or event. For example, if the average attendance is 12, it should be displayed on top of the progress bar instead of in a new column.
- Verify that the capacity and registered students are shown in one column for each tutorial group or event. For example, 10/20 should be displayed instead of separate columns for capacity and registered students.
- Verify that the language, campus, and online/offline columns are hidden if they are equal for all tutorial groups. For example, if all tutorial groups or events have the same language (e.g. English), campus (e.g. Main Campus) and online/offline status (e.g. Online), these columns should not be displayed at all.

### Review Progress


#### Code Review
#### Manual Tests
### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| tutorial-groups-table.component.ts | 79% | 91% |
| tutorial-group-row.component.ts| 100% | 100% |
| remove-seconds.pipe.ts | 100% | 100% |
| meeting-pattern.pipe.ts | 84% | 95% |


### Screenshots

#### Before Instructor View:
<img width="1015" alt="instructor_view_before" src="https://user-images.githubusercontent.com/29718932/226187648-058b1a7a-aca7-487d-8666-513228455a14.png">

#### Before Student View:
![student_view_all_before](https://user-images.githubusercontent.com/29718932/226187662-634721cb-aded-47bf-8f13-c93b36e0334f.png)

#### Instructor After:
![management_view_new](https://user-images.githubusercontent.com/29718932/226215862-180c6d5a-5bf8-476f-b6d5-39f39f794b8f.png)
Note: **undefined bug is already fixed**

Same Language, Campus, Online,Offline:

![SCR-20230320-bfif](https://user-images.githubusercontent.com/29718932/226215909-d7edd899-c50d-4e2f-a7a0-f121f19d9744.png)


#### Student After:
![management_view_all_same](https://user-images.githubusercontent.com/29718932/226215866-73b7dbc2-659a-4d61-846e-ae3cbe7ce6e0.png)
